PYIC-1511 Updates to the dev env deployment script

vam-gds opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed
Updates to the dev env deployment script

Further updates to script, fixed all remaining actual issues

	modified:   utils/README.md
	modified:   utils/deploy_to_dev_env.sh
<!-- Describe the changes in detail - the "what"-->

### Why did it change
This is just some quality of life improvements to the deployment shell script. 

<!-- Describe the reason these changes were made - the "why" -->

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-1511](https://govukverify.atlassian.net/browse/PYIC-1511)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [ ] No environment variables or secrets were added or changed

<!-- Delete if changes DO NOT include new environment variables or secrets -->
- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

vam-gds committed:
PYIC-1511 Updates to the dev env deployment script

Further updates to script, fixed all remaining actual issues

	modified:   utils/README.md
	modified:   utils/deploy_to_dev_env.sh

danworth commented on utils/deploy_to_dev_env.sh:
```diff
@@ -4,55 +4,120 @@
-aws ecr get-login-password --region eu-west-2 | \
-  docker login --username AWS --password-stdin "${ecr_registry}"
+function clone_config_repo() {
+  git clone git@github.com:alphagov/di-ipv-config.git config-repo
```
So at present if I do not provide the `-e` flag the error message isn't right because process_args is called before init and the `environment` array is not populated, not a huge deal but I imagine it'll happen frequently the first time people run it. This is what you get
```
GDS11321:utils dan.worth$ aws-vault exec di-ipv-dev -- ./deploy_to_dev_env.sh

Usage: ./deploy_to_dev_env.sh:
        [-e] Provide environment name from:
```


I get why you're cloning the repo but ideally I don't want to wait to clone it when I already have it (its also unexpected). For simplicity why not just add a link to the file in the help text, something like the following and add anything extra to the README? Or look for it on the dev's laptop, most people will clone it into the same parent directory. I feel like we're on the brink of over engineering this, despite my fondness for a bit of `bash` :)

```
echo -e "\t[-e] Provide environment name from those listed at https://github.com/alphagov/di-ipv-config/blob/main/core/ci/core-developer-pipelines/generate-pipelines/list_of_developers.txt\n"
```

danworth replied:
If you want to be helpful and provide a list of configured envs then how about using something like the following to derive it from the available front end stacks?? This also means it'll make sure the stack actually exists which is necessary further on....
```
GDS11321:utils dan.worth$ aws-vault exec dev-admin -- aws cloudformation list-stacks --region eu-west-2 | jq '.StackSummaries[] | select(.StackName | contains("front")).StackName' -r | sort | uniq | cut -d '-' -f 3,4
dev-amrits
dev-aprilm
dev-chrisw
dev-danh
dev-danp
dev-danw
dev-lait
dev-patrickb
dev-vamh
dev-wojciecho
```

danworth replied:
To be honest I'm leaning towards a link to the file and a note that if they're not listed to ask for their environment to be set up. That seems the most simple but open to other ideas.

vam-gds replied:
so I have pushed up updates which sorts this out. 

We cant be sure they have the correct credentials that early on so aws cant be used. 

I have completely reworked the git clone it downloads only 448K:

```
du -sh config-repo/
488K    config-repo/
```

it runs before anything else quietly downloads a very sparse version of the repo assigns the contents of that file to an array and then removes itself and everything now works as expected (quietly in the background no one will be aware minimal interfernace) 

I do this because I cant be sure they are calling aws-vault or gds to correct env, where the ipv-config repo resides what branch it may be on, when they last updated it etc - this way its completely clean we get the latest update directly from what should be a source of truth and it should be very quick now considering its less than half a meg download. 

so for me locally:

```
$ time aws-vault exec core-dev -- ./deploy_to_dev_env.sh

Usage: ./deploy_to_dev_env.sh:
        [-e] Provide environment name from: dev-danw dev-danp dev-danh dev-chrisw dev-amrits dev-kerrr dev-lait dev-ianj dev-derrenw dev-patrickb dev-vamh dev-wojciecho dev-aprilm

        [-h] Print this usage guide


real    0m1.814s
user    0m0.085s
sys     0m0.055s
```
```
$ time aws-vault exec core-dev -- ./deploy_to_dev_env.sh -e test
        ./deploy_to_dev_env.sh: Error: 'test' does not match any of the valid known environments: dev-danw dev-danp dev-danh dev-chrisw dev-amrits dev-kerrr dev-lait dev-ianj dev-derrenw dev-patrickb dev-vamh dev-wojciecho dev-aprilm


Usage: ./deploy_to_dev_env.sh:
        [-e] Provide environment name from: dev-danw dev-danp dev-danh dev-chrisw dev-amrits dev-kerrr dev-lait dev-ianj dev-derrenw dev-patrickb dev-vamh dev-wojciecho dev-aprilm

        [-h] Print this usage guide


real    0m1.947s
user    0m0.085s
sys     0m0.067s
```

roughly about 2 seconds now.

danworth commented on utils/deploy_to_dev_env.sh:
```diff
@@ -61,27 +126,35 @@ function isStackUpdateComplete() {
 
   echo "Current status is: ${status}"
-  if [[ "$status" == "UPDATE_COMPLETE" ]]; then
+  if [[ " ${stack_status[*]} " =~ " ${status} " ]]; then
```
ah nice, haven't seen this done in a while :)

vam-gds committed:
Fixed loopholes around specific conditions

Completely reworked the git cloning it now does a tiny clone of the
bare essentials and cleans itself up afterwards - all issues are resolved now

fixed supported flags

vam-gds committed:
PYIC-1511 Added new deploy_dev script

This works in a silightly different fashion it keeps the config-repo and
instead runs a git pull on each invocation to keep itself constantly
updated.

I have a fast connection so for me

deploy_dev first run range 3.2 - 3.8s consequent runs 1.4s - 1.6s

deploy_to_dev_env.sh always around 1.7 - 2s

for me not much difference (I prefer the deploy_to_dev_env as it is
cleaner imho), however for those on a slower connection maybe the
deploy_dev method will be better initially it will be a bit slower
but consequent runs should be faster.

I have uploaded both methods to see which one preforms better, for
others or which is preferred.

	new file:   deploy_dev.sh

vam-gds committed:
PYIC-1511 removed environment checking

danworth commented on utils/deploy_to_dev_env.sh:
```diff
@@ -4,55 +4,104 @@
 
+function check_dependencies() {
+  if ! which jq >/dev/null; then
+    print_error_exit "Please install jq to use this script - more info here: https://formulae.brew.sh/formula/jq"
```
this is a helpful addition 👍

danworth commented on utils/deploy_to_dev_env.sh:
```diff
@@ -4,55 +4,104 @@
+  echo -e "\t* Stack Name:  '${STACK_NAME}'"
+  echo -e "\t* Image Tag:   '${DEV_IMAGE_TAG}'\n"
+  check_dependencies
+  check_connection
```
Useful summary 👍

danworth reviewed (approved):
🏅
